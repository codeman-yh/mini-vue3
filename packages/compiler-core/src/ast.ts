export const enum NodeTypes{
    ROOT,
    INTERPOLATION,
    ELEMENT,
    TEXT,
    SIMPLE_EXPRESSION,
    COMPOUND_EXPRESSION
}