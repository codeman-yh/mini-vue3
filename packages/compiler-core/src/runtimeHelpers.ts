export const TODISPLAYSTRING = Symbol('toDisplayString')
export const CREATEELEMENTBLOCK = Symbol('createElementBlock')

export const helperNameMap = {
    [TODISPLAYSTRING]: 'toDisplayString',
    [CREATEELEMENTBLOCK]:'createElementBlock'
}