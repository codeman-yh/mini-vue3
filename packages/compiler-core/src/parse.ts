import { NodeTypes } from "./ast"

const enum TagType{
    start,
    end
}

export function baseParse(content: string){
    const context = creatParserContext(content)
    return createRoot(parseChildren(context,[]))
}

function parseChildren(context,ancestors){
    const nodes:any = []
    while(!isEnd(context,ancestors)){
        const s = context.source
        let node;
        if(s.startsWith("{{")){
            node = parseInterpolation(context)
        }else if(s[0] == "<"){
            if(/^[a-z]/i.test(s[1])){
                node = parseElement(context,ancestors)
            }
        }else{
            node = parseText(context)
        }
        nodes.push(node)
    }
    return nodes
}



function parseInterpolation(context){
    const openDelimiter = "{{"
    const closeDelimiter = "}}"

    let closeIndex = context.source.indexOf(closeDelimiter,openDelimiter)

    advanceBy(context,openDelimiter.length)
    let rawContent =  parseTextData(context, closeIndex - closeDelimiter.length)
    let content = rawContent.trim()
    advanceBy(context,closeDelimiter.length)

    return {      
        type: NodeTypes.INTERPOLATION,
        content: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content
        }       
    }   
}

function parseElement(context,ancestors){
    const elementNode:any = parseTag(context,TagType.start)
    ancestors.push(elementNode)
    elementNode.children = parseChildren(context,ancestors)
    ancestors.pop()
    if(isCloseWithSameTag(context.source,elementNode.tag)){
        parseTag(context,TagType.end)
    }else{
        throw new Error('缺失结束标签：span')
    }
    return elementNode
}

function parseTag(context, tagType){
    let match = context.source.match(/^<\/?([a-z]*)/i)
    let tag = match![1]
    advanceBy(context, match[0].length)
    advanceBy(context,1)
    if(tagType == TagType.start){
        return{       
            type: NodeTypes.ELEMENT,
            tag,
        }
    }

}

function parseText(context){
    const s = context.source
    const endTokens = ["{{","<"]
    let endIndex = s.length;
    for(let i=0; i<endTokens.length;i++){
        const index  = s.indexOf(endTokens[i])
        if(index !== -1 && index < endIndex){
            endIndex = index
        }
    }
    const content = parseTextData(context,endIndex)
    return {
        type: NodeTypes.TEXT,
        content,
    }
}

function parseTextData(context,length){
    const content = context.source.slice(0,length)
    advanceBy(context,length)
    return content
}

function creatParserContext(content: string) {
    return{
        source: content
    }
}
function createRoot(children) {
    return{
        type: NodeTypes.ROOT,
        children,
        helpers: [],
    }
}

function advanceBy(context, length) {
    context.source = context.source.slice(length)
}

function isEnd(context,ancestors){
    const s = context.source
    // 还需判断节点是否需要闭合
    if(s.startsWith("</")){
        for(let i= ancestors.length -1;i>=0;i--){
            if(isCloseWithSameTag(s,ancestors[i].tag)){
                return true
            }
        }
        return 
    }
    return !context.source
}

function isCloseWithSameTag(s,tag){
    return s.slice(2,tag.length + 2) === tag
}
