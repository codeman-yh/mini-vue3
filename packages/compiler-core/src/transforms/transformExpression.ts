import { NodeTypes } from "../ast";

export function transformExpression(node){
    if(node.type === NodeTypes.INTERPOLATION){
        processExpression(node.content)
    }
}

function processExpression(contentNode) {
    contentNode.content = `_ctx.${contentNode.content}`
}
