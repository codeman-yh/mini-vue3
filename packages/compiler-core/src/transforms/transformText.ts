import { NodeTypes } from "../ast";

export function transformText(node){
    if(node.type === NodeTypes.ELEMENT){
        for(let i=0; i<node.children.length;i++){
            let currentContainer:any = {};
            let child = node.children[i]
            // 查看下一个是否为Text类型的节点
            if(isText(child)){
                for(let j=i+1; j<node.children.length; j++){
                    let next = node.children[j]
                    if(isText(next)){
                        if(!currentContainer.children){
                           // 创建初始值
                           currentContainer = node.children[i] = {
                                type: NodeTypes.COMPOUND_EXPRESSION,
                                children:[child]
                           }
                        }
                        currentContainer.children.push(" + ", next)                       
                        // 将添加进COMPOUND类型的数据从原数组删除
                        node.children.splice(j,1)
                        j--;
                    }else{
                        break;
                    }
                }  
            }
    
        }
    }
}

function isText(node){
   return node.type === NodeTypes.TEXT || node.type === NodeTypes.INTERPOLATION
}