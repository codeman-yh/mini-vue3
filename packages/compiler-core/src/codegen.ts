import { isString } from "@mini-vue3/shared"
import { NodeTypes } from "./ast"
import { CREATEELEMENTBLOCK, helperNameMap, TODISPLAYSTRING } from "./runtimeHelpers"

export function generate(ast){
    const context = createGenerateContext(ast)
    const {push} = context

    genFunctionPreamble(ast, context)

    const functionName = 'render'
    const arg = ['_ctx']


    push(`return function ${functionName}`)
    push(`(${[...arg]}){`)
    push(`return `)

    genNode(ast.codegenNode, context);

    push("}")
    return context

}

function genNode(node: any, context) {
    switch (node.type){
        case NodeTypes.TEXT:
            genText(node,context)
            break;
        case NodeTypes.INTERPOLATION:
            genInterpolation(node,context)
            break;
        case NodeTypes.SIMPLE_EXPRESSION:
            genExpression(node,context)
            break;
        case NodeTypes.ELEMENT:
            genElement(node,context)
            break;
        case NodeTypes.COMPOUND_EXPRESSION:
            genCompoundExpression(node, context);
            break;
    }
}

function genText(node: any, context: any){
    const {push} = context
    push(`'${node.content}'`)
}

function genInterpolation(node: any, context: any) {
    const {push} = context
    push(`${context.helper(TODISPLAYSTRING)}(`)
    genNode(node.content,context)
    push(")")
}

function genExpression(node: any, context: any){
    context.push(`${node.content}`)
}

function genElement(node: any,context: any){
    const {push} = context
    const {tag, props, children} = node
    push(`${context.helper(CREATEELEMENTBLOCK)}(`)
    genNodeList(genNullableArgs([tag,props,children]),context)
    push(")")
}

function genCompoundExpression(node: any, context: any) {
    const {push} = context
    for(let i=0; i<node.children.length; i++){
        if(isString(node.children[i])){
            push(node.children[i])
        }else{ 
            genNode(node.children[i],context)
        }
    }
}

function genNodeList(nodes,context){
    const {push} = context
    for(let i=0;i<nodes.length;i++){
        const node = nodes[i]
        if(isString(node)){
            push(`${node}`)
        }else{
            genNode(node,context)
        }
        if (i < nodes.length - 1) {
            push(", ");
          }

    }
}

function genNullableArgs(args){
    return args.map(arg => arg || 'null')
}

function createGenerateContext(ast: any) {
    const context = {
        code:'',
        push(code){
            context.code += code
        },
        helper(key) {
            return `_${helperNameMap[key]}`;
        },
    }
    return context
}



function genFunctionPreamble(ast, context) {
   const {push} = context
   const name = 'Vue'
   const aliasHelper = (s) => `${helperNameMap[s]} : _${helperNameMap[s]}`
   if(ast.helpers.length > 0){
    const arg = ast.helpers.map(aliasHelper).join(",")
    push(`const { ${arg} } = ${name} `)
   }
   push('\n')
}


