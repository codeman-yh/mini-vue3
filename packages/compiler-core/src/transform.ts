import { NodeTypes } from "./ast";
import { TODISPLAYSTRING } from "./runtimeHelpers";

export function transform(root,options:any={}){
    const context = createTransformContext(root,options)
    // 遍历ast
    traverseNode(root,context)
    createRootCodegen(root)
    root.helpers.push(...context.helpers.keys());
}

function traverseNode(node,context) {
    const type:NodeTypes = node.type
    let plugins = context.nodeTransfrom


    switch (type){
        case NodeTypes.INTERPOLATION:
            context.helper(TODISPLAYSTRING)
        break;
        case NodeTypes.ROOT:
        case NodeTypes.ELEMENT:
            traverseChildren(node,context)
        break;
    }

    // 这里执行是保存闭包的节点？
    if(plugins){
        plugins.forEach(plugin => {
            plugin(node,context)
        });
    }
}

function traverseChildren(node,context){
    let astChildren = node.children
    if(astChildren){
        for(let i=0; i<astChildren.length;i++){
            traverseNode(astChildren[i],context)            
        }
    }
}

function createTransformContext(root: any, options: any) {
    const context = {
        root,
        nodeTransfrom: options.nodeTransfrom || [],
        helpers: new Map(),
        helper(name){
            const count = context.helpers.get(name) || 0
            context.helpers.set(name,count + 1)
        }
    }
    return context
}

function createRootCodegen(root) {
    const {children} = root
    const node = children[0]
    if(node.codegenNode){
        root.codegenNode = node.codegenNode
    }else{
        root.codegenNode = node
    }

}


