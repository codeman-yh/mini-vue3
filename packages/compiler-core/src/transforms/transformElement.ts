import { NodeTypes } from "../ast"
import { CREATEELEMENTBLOCK } from "../runtimeHelpers"

export function transformElement(node,context){
    if(node.type === NodeTypes.ELEMENT){
        context.helper(CREATEELEMENTBLOCK)
        const childNode = node.children[0]
        const vTag = `'${node.tag}'`
        const vProps = node.props
        const vnode = {
            type:NodeTypes.ELEMENT,
            tag: vTag,
            props: vProps,
            children: childNode
        }
        node.codegenNode = vnode
    }

}