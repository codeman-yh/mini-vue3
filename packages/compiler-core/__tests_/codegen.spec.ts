import { generate } from "../src/codegen"
import { baseParse } from "../src/parse"
import { transform } from "../src/transform"
import { transformElement } from "../src/transforms/transformElement"
import { transformExpression } from "../src/transforms/transformExpression"
import { transformText } from "../src/transforms/transformText"

describe('codegen',()=>{
    test('text',()=>{ 
        const ast = baseParse('hello')
        transform(ast)
        const {code} = generate(ast)
        expect(code).toMatchSnapshot()
    })

    test('interploation',()=>{
        const ast = baseParse('{{message}}')
        transform(ast,{
            nodeTransfrom: [transformExpression]
        })
        const {code} = generate(ast)
        expect(code).toMatchSnapshot()
    })

    test('element',()=>{
        const ast = baseParse('<div>hiï¼Œ{{message}}</div>')
        transform(ast,{
            nodeTransfrom:[transformExpression,transformText,transformElement]
        })
        console.log(ast)
        const {code} = generate(ast)
        expect(code).toMatchSnapshot()
    })
})