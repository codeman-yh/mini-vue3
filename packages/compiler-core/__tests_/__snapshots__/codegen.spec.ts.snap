// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen element 1`] = `
"const { toDisplayString : _toDisplayString,createElementBlock : _createElementBlock } = Vue 
return function render(_ctx){return _createElementBlock('div', null, 'hiï¼Œ' + _toDisplayString(_ctx.message))}"
`;

exports[`codegen interploation 1`] = `
"const { toDisplayString : _toDisplayString } = Vue 
return function render(_ctx){return _toDisplayString(_ctx.message)}"
`;

exports[`codegen text 1`] = `
"
return function render(_ctx){return 'hello'}"
`;
